lib/seahorn/PredicateAbstraction.cc

```diff
@@ -0,0 +1,582 @@
+
+					new_body_exprs.push_back(equal_expr);
+					index ++;
+				}
+				converter.addRelToBoolToTerm(bind::fname(rule_head), boolToTermMap);
+				//converter.getRelToBoolToTermMap().insert(std::pair<Expr, ExprMap>(bind::fname(rule_head), boolToTermMap));
+			}
+
+			//Extract the constraints
+			Expr constraints = util_extractTransitionRelation(r, db);
+			new_body_exprs.push_back(constraints);
+
+			//Construct new body
+			Expr new_rule_body = mknary<AND>(new_body_exprs.begin(), new_body_exprs.end());
+			LOG("pabs-debug", outs() << "NEW RULE BODY :" << *new_rule_body << "\n";);
+
+			HornRule new_rule(rule_vars, new_rule_head, new_rule_body);
+			new_DB.addRule(new_rule);
+		}
+	}
+
+	void PredicateAbstractionAnalysis::generateAbstractQueries(HornClauseDB &db, HornClauseDB &new_DB)
+	{
+		for(auto old_query : db.getQueries())
+		{
+			Expr old_fdecl = bind::fname(old_query);
+			Expr new_fdecl = m_oldToNewPredMap.find(old_fdecl)->second;
+			ExprVector old_args;
+			Expr new_query = bind::fapp(new_fdecl, old_args);
+			new_DB.addQuery(new_query);
+		}
+	}
+
+	void PredicateAbstractionAnalysis::guessCandidate(HornClauseDB &db)
+	{
+	  for(Expr rel : db.getRelations())
+	  {
+		  if(bind::isFdecl(rel))
+		  {
+			  //ExprVector terms = relToCand(rel);
+			  ExprVector terms = applyTemplatesFromExperimentFile(rel, "/home/chenguang/Desktop/seahorn/test/pabs-experiment/preds_temp");
+			  m_currentCandidates.insert(std::pair<Expr, ExprVector>(rel, terms));
+		  }
+	  }
+	}
+
+	ExprVector PredicateAbstractionAnalysis::applyTemplatesFromExperimentFile(Expr fdecl, std::string filepath)
+	{
+		ExprVector bvars;
+		ExprVector lemmas;
+
+		int bvar_count = 0;
+		for(int i=0; i<bind::domainSz(fdecl); i++)
+		{
+			if(isOpX<INT_TY>(bind::domainTy(fdecl, i)))
+			{
+				Expr bvar = bind::bvar(i, mk<INT_TY>(bind::domainTy(fdecl, i)->efac()));
+				bvars.push_back(bvar);
+				bvar_count ++;
+			}
+		}
+		Expr one = mkTerm<mpz_class> (1, fdecl->efac());
+		Expr zero = mkTerm<mpz_class> (0, fdecl->efac());
+		Expr two = mkTerm<mpz_class> (2, fdecl->efac());
+		if(bvar_count == 0)
+		{
+			lemmas.push_back(mk<TRUE>(fdecl->efac()));
+		}
+		else if(bvar_count == 1)
+		{
+			lemmas.push_back(mk<GEQ>(bvars[0], one));
+			lemmas.push_back(mk<LEQ>(bvars[0], one));
+			lemmas.push_back(mk<GEQ>(bvars[0], zero));
+			lemmas.push_back(mk<GEQ>(bvars[0], two));
+			lemmas.push_back(mk<LEQ>(bvars[0], two));
+			parseLemmasFromExpFile(bvars[0], lemmas, filepath);
+		}
+		else
+		{
+			for(int i=0; i<bvars.size(); i++)
+			{
+				lemmas.push_back(mk<GEQ>(bvars[i], one));
+				lemmas.push_back(mk<LEQ>(bvars[i], one));
+				lemmas.push_back(mk<GEQ>(bvars[i], zero));
+				lemmas.push_back(mk<GEQ>(bvars[i], two));
+				lemmas.push_back(mk<LEQ>(bvars[i], two));
+				parseLemmasFromExpFile(bvars[i], lemmas, filepath);
+			}
+		}
+		return lemmas;
+	}
+
+	void PredicateAbstractionAnalysis::parseLemmasFromExpFile(Expr bvar, ExprVector& lemmas, std::string filepath)
```

Comment:
Is it possible to store lemmas in some smt-lib compatible representation and use smt-lib parser instead of a custom one?